cpp/src/statistics.cpp

```diff
@@ -81,6 +81,12 @@ std::vector<std::string> Statistics::list_stat_names() const {
     return std::vector<std::string>{ks.begin(), ks.end()};
 }
 
+void Statistics::clear() {
+    std::lock_guard<std::mutex> lock(mutex_);
+    stats_.clear();
+    memory_records_.clear();
+}
```

Comment:
What happens if we are in the middle of a `MemoryRecorder`?
I think we need to throw an error if `memory_records_` isn't empty otherwise `Statistics::MemoryRecorder::~MemoryRecorder()` will fail.

Reply:
I've done statistics only for now.

Reply:
> I'm not familiar with `MemoryRecorder`. My hope was that the lock we acquire here in `clear` would prevent us from modifying things when we aren't supposed to.

When memory profiling is enabled, a `MemoryRecorder` register a record on creation and tally it up on destruction.